[CMake] Allow enabling C++ library assertions if supported

aperezdc opened the pull request:
#### 3b5d8ebc05c0985537639b9eaa92ab722607a2fa
<pre>
[CMake] Allow enabling C++ library assertions if supported
<a href="https://bugs.webkit.org/show_bug.cgi?id=266396">https://bugs.webkit.org/show_bug.cgi?id=266396</a>

Reviewed by Michael Catanzaro.

Detect which C++ library is being used, and for the supported ones apply
the needed compilation options to have their assertions enabled. The
_LIBCPP_VERSION macro is used to detect libc++, and _GLIBCXX_RELEASE for
GNU libstdc++. If any of these C++ libraries are detected, then their
assertions are enabled by default. A new USE_CXX_STDLIB_ASSERTIONS CMake
option may be used used to override the default setting.

This change does not attempt to choose whether to use libc++, and the
mechanism to choose the C++ library being used remains unchanged. To
change e.g. from the GNU library to libc++ one still needs to set
environment variables accordingly before invoking CMake:

  CXX=clang++ CXXFLAGS=&apos;-stdlib=libc++&apos; LDFLAGS=&apos;-stdlib=libc++&apos; cmake ...

* Source/cmake/OptionsCommon.cmake: Add checks to determine which C++
standard library is being used and enable assertions where possible.

Canonical link: <a href="https://commits.webkit.org/290394@main">https://commits.webkit.org/290394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0839d8b4bf494ff09a436a5361f90523a0450b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89563 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94555 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40330 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17369 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69004 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26638 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7012 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39436 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82361 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77366 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36676 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96383 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88338 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16745 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12296 "Found 3 new test failures: http/tests/pdf/linearized-pdf-in-display-none-iframe.html imported/w3c/web-platform-tests/css/selectors/invalidation/modal-pseudo-class-in-has.html pdf/pdf-in-styled-embed.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77870 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17000 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77186 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21605 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20192 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9912 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14116 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16758 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110831 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16499 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26568 "Found 4 new JSC stress test failures: wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19950 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
This is a good change. Well, sort of. Some background info:

 * Both Fedora and freedesktop-sdk build with `-Wp,-D_GLIBCXX_ASSERTIONS` so assertions should be enabled already
 * Fedora builds WebKitGTK with GCC. GNOME builds with Clang since it's faster. But both use libstdc++, not libc++

So the reason for my "sort of" is that I would class this as a hardening flag, and we currently do not set any of the standard hardening flags that are required for WebKit to be actually safe to use, and it's _slightly_ weird to set just one particular such flag and not others. Distributors are expected to set them manually and users suffer an extremely severe security penalty if they do not. Maybe we should reconsider this approach and set all standard hardening flags ourselves, even though no other software does this. (For whatever reason, it's just expected that distributors and not software projects control the hardening flags. But maybe change starts here?) I'm OK with this anyway, because setting just stdlib assertions is a net benefit. And it's also a little different from the other normal hardening flags because it facilitates debugging as well; I'd be pretty surprised if this change doesn't "break" any layout tests by exposing bugs, for example (but our EWS won't notice because they don't use libc++).

I think it's safe to enable assertions by default for libstdc++ too. Red Hat's toolchain team is extremely conservative (you've seen their pushback against frame pointers over 1-2% overhead) and if they think it's safe to build the entire OS with the assertions enabled, it's probably not going to make a big performance difference.

The one minor downside to adding this ourselves is the flag will now be duplicated in distro builds when the distro enables it already. That's probably OK.

(Anyone looking for documentation of the flags used by Fedora, which freedesktop-sdk mostly follows, should read [this](https://src.fedoraproject.org/rpms/redhat-rpm-config//blob/rawhide/f/buildflags.md).)

mcatanzaro reviewed (commented):
So this is surely a good change, but I'm afraid the WPE EWS don't like it. This isn't too surprising because WPE doesn't receive the same degree of downstream distro testing that WebKitGTK does, so it probably hasn't been run with stdlib assertions before.

aperezdc committed:
[CMake] Allow enabling C++ library assertions if supported
https://bugs.webkit.org/show_bug.cgi?id=266396

Reviewed by Michael Catanzaro.

Detect which C++ library is being used, and for the supported ones apply
the needed compilation options to have their assertions enabled. The
_LIBCPP_VERSION macro is used to detect libc++, and _GLIBCXX_RELEASE for
GNU libstdc++. If any of these C++ libraries are detected, then their
assertions are enabled by default. A new USE_CXX_STDLIB_ASSERTIONS CMake
option may be used used to override the default setting.

This change does not attempt to choose whether to use libc++, and the
mechanism to choose the C++ library being used remains unchanged. To
change e.g. from the GNU library to libc++ one still needs to set
environment variables accordingly before invoking CMake:

  CXX=clang++ CXXFLAGS='-stdlib=libc++' LDFLAGS='-stdlib=libc++' cmake ...

* Source/cmake/OptionsCommon.cmake: Add checks to determine which C++
standard library is being used and enable assertions where possible.

Canonical link: https://commits.webkit.org/290394@main
